### PR TITLE
Closes #19846: Show Default Browser Notification if browser is not default

### LINKED ISSUE #19846: Show 'Default Browser' Local Notification for New users
Built as a feature of the app, (much like the onboarding flow) that gets scheduled in a user's first session and triggers if they do not set us as default 24 hours after 1st app launch, see an example below.

image

### PR DESCRIPTION
**Pull request by rocketsroger**
https://user-images.githubusercontent.com/6579541/123160851-a9fd2d00-d43c-11eb-8074-56355dfb1545.mp4

### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by rocketsroger** [comment]
@topotropic can you please confirm if the notification text is correct.  Currently it is "Make <app name> your default browser"

**Comment by topotropic** [comment]
This looks good! Thanks

**Review by topotropic — APPROVED** [review]

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    override fun doWork(): Result {
+        ensureChannelExists()
+        NotificationManagerCompat.from(applicationContext)
+            .notify(NOTIFICATION_TAG, NOTIFICATION_ID, buildNotification())
+        return Result.success()
+    }
+
+    /**
+     * Build the notification with site controls to be displayed while the web app is active.
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:39]: What does this have to do with web apps? 🤔

File: app/src/main/java/org/mozilla/fenix/home/intent/DefaultBrowserIntentProcessor.kt [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import org.mozilla.fenix.onboarding.DefaultBrowserNotificationWorker.Companion.isDefaultBrowserNotificationIntent
+
+class DefaultBrowserIntentProcessor(
+    private val activity: HomeActivity
+) : HomeIntentProcessor {
+
+    override fun process(intent: Intent, navController: NavController, out: Intent): Boolean {
+        return if (isDefaultBrowserNotificationIntent(intent)) {
+            activity.openSetDefaultBrowserOption()
+            activity.settings().shownDefaultBrowserNotification = true
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/home/intent/DefaultBrowserIntentProcessor.kt:21]: hand-wavy nit: I was misreading "shown" as "show" so the logic was a bit confusing at first. We have some examples of "should" but that's the opposite logic that you have here. Maybe a good past tense word could be "defaultBrowserNotificationDisplayed"/"showedDefaultBrowserNotification'?

Again, this is a pretty handy-wavy nit, so by all means, ignore it if you wish.

File: app/src/main/java/org/mozilla/fenix/utils/Settings.kt [hunk 3]
```diff
@@ -479,6 +479,15 @@ class Settings(private val appContext: Context) : PreferencesHolder {
         return browsers.isDefaultBrowser
     }
 
+    var shownDefaultBrowserNotification by booleanPreference(
+        appContext.getPreferenceKey(R.string.pref_key_should_show_default_browser_notification),
+        default = false
+    )
+
+    fun shouldShowDefaultBrowserNotification(): Boolean {
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/utils/Settings.kt:487]: nit: Maybe move this to where ever it's only use case is. It doesn't seem like this function needs to be globally accessible from all of settings. It's a strange pattern that seems to exist in this class for some reason. 🤷

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 4]
```diff
@@ -338,6 +345,8 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
 
             settings().wasDefaultBrowserOnLastResume = settings().isDefaultBrowser()
         }
+
+        setDefaultBrowserNotificationIfNeeded()
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/HomeActivity.kt:349]: This touches IO via SharedPreferences for every resume. We may want to take some care around adding this.

cc: @mcomella

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 5]
```diff
@@ -158,7 +164,8 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
             SpeechProcessingIntentProcessor(this, components.core.store, components.analytics.metrics),
             StartSearchIntentProcessor(components.analytics.metrics),
             OpenBrowserIntentProcessor(this, ::getIntentSessionId),
-            OpenSpecificTabIntentProcessor(this)
+            OpenSpecificTabIntentProcessor(this),
+            DefaultBrowserIntentProcessor(this)
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/HomeActivity.kt:168]: Looking at the implementation of `DefaultBrowserIntentProcessor`, I don't see this touching IO but our intent processors increase, so our VIEW startup performance might be affected.

cc: @mcomella

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                .setColor(ContextCompat.getColor(this, R.color.primary_text_light_theme))
+                .setPriority(NotificationCompat.PRIORITY_LOW)
+                .setShowWhen(false)
+                .setContentIntent(pendingIntent)
+                .build()
+        }
+    }
+
+    /**
+     * Make sure a notification channel for the powered by notifications exists.
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:69]: Docs from a different place?

Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:69]: Seems like this class is coming from `PoweredByNotification`. Is there anything around here that would be good to abstract out for re-use?

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+     * Returns the channel id to be used for notifications.
+     */
+    private fun ensureChannelExists(): String {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            val notificationManager: NotificationManager =
+                applicationContext.getSystemService(NOTIFICATION_SERVICE) as NotificationManager
+
+            val channel = NotificationChannel(
+                NOTIFICATION_CHANNEL_ID,
+                applicationContext.getString(R.string.notification_default_browser_channel_name),
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:80]: Is it intentional to have a notification channel _only_ for the default browser notification or is there something more generic if/when we want to have more that fall under that grouping? I can see Brave has a generic "Brave Ads" [group ID](https://github.com/brave/brave-core/blob/1cb5818aa0b70666c6aeea5ea9c06cc4e712171a/android/java/org/chromium/chrome/browser/notifications/channels/BraveChannelDefinitions.java#L19).

**Review by jonalmeida — COMMENTED** [review]
Left some drive-by comments. 🚗 🙂

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 4 above)
Comment (mcomella) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:349]: Thanks for catching this. Can we do this on a background thread? afaict, there's nothing here that requires UI thread interaction.

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            return NotificationCompat.Builder(this, channelId)
+                .setSmallIcon(R.drawable.ic_status_logo)
+                .setContentTitle(
+                    applicationContext.getString(R.string.notification_default_browser_text, appName))
+                .setBadgeIconType(BADGE_ICON_NONE)
+                .setColor(ContextCompat.getColor(this, R.color.primary_text_light_theme))
+                .setPriority(NotificationCompat.PRIORITY_LOW)
+                .setShowWhen(false)
+                .setContentIntent(pendingIntent)
+                .build()
```
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:64]: Is `PRIORITY_LOW` intentional? That means, by your video as well, that the notification will never "pop-up" when it is delivered but it will hide (minimized) until the user opens up their either notification tray and find it there. That seems like the opposite desire for a notification that we want to catch the user's attention with. We might want to ask UX or marketing about that.

With `BADGE_ICON_NONE` might have similar effects of hiding the notification, but I'm not fully sure of what [the docs](https://developer.android.com/reference/android/app/Notification?hl=ja#BADGE_ICON_NONE) mean.

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt (hunk 1 above)
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:39]: my mistake.  forgot to update the comments.

File: app/src/main/java/org/mozilla/fenix/home/intent/DefaultBrowserIntentProcessor.kt (hunk 2 above)
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/home/intent/DefaultBrowserIntentProcessor.kt:21]: Sure, I'll rename to `defaultBrowserNotificationDisplayed`.  Thanks,

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 4 above)
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:349]: Agreed.  Thanks for catching this.  I'll do this on a background thread.

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 5 above)
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:168]: The processing of the intent is fairly quick.  I don't think it will affect the VIEW startup performance unless the intent is from the default browser notification.  It checks if the intents have the `INTENT_DEFAULT_BROWSER_NOTIFICATION` extra.  if not then it returns right away.

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt (hunk 7 above)
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:80]: Sure, I'll update the channel name to `Marketing` to allow for more generic usage in the future.  Thanks,

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt (hunk 6 above)
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:69]: Yes, this might be a good abstraction for the support library.   Currently there's a lot of `ensureChannelExists` function calls in both Fenix and Android Components.  I'll created an issue to track this work.

Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:69]: https://github.com/mozilla-mobile/android-components/issues/10520

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt (hunk 8 above)
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:64]: This is a good question.  My assumption is that we don't want to make any noise with this notification.  However, this is just my assumption.  @topotropic, do we want this notification more visible to the user?  Thanks,

Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:64]: I'll update the priority to `PRIORITY_DEFAULT` and `BADGE_ICON_SMALL` so it's more visible.  I'll upload a new recording after.  Thanks,

**Comment by rocketsroger** [comment]
https://user-images.githubusercontent.com/6579541/123292759-5db7f880-d4e1-11eb-95e0-baafe2d0e361.mp4

Updated behaviour.

File: app/src/main/java/org/mozilla/fenix/utils/Settings.kt (hunk 3 above)
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/utils/Settings.kt:492]: Sounds good.  I'll move it thanks,

Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/utils/Settings.kt:492]: Hmm, this makes testing more complicated.  I think I'll leave this where it is.  I would like to be able to test this without other dependencies.

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 5 above)
Comment (mcomella) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:168]: In the event that we take the `true` branch of the if statement in that class, I'm also concerned this could impact start up perf:
```kotlin
        return if (isDefaultBrowserNotificationIntent(intent)) {
            activity.openSetDefaultBrowserOption()
            activity.settings().defaultBrowserNotificationDisplayed = true
```

- `openSetDefaultBrowserOption` appears to access a `SystemService`. The `SystemService` classes often rely on IPC which can be a little slow (e.g. this is why we're modifying how `isDefaultBrowser` is called in https://github.com/mozilla-mobile/fenix/pull/20116/)
- We write to settings and any access of shared prefs is has some perf impact

Perhaps you can measure how long this branch takes _on a low-end device_ like the Moto G5? e.g.
```kotlin
val before = SystemClock.elapsedRealtime()
return if (isDefaultBrowserNotificationIntent(intent)) {
            activity.openSetDefaultBrowserOption()
            activity.settings().defaultBrowserNotificationDisplayed = true
            true
...
}.also {
    val end = SystemClock.elapsedRealtime()
    Log.e("lol", "elapsed: ${end - start}")
}
```

If you don't have a low-end device, let me know and I can take the measurement if you send me a build. :)

Comment (mcomella) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:168]: If the timing is 10ms+ on a Moto G5-equivalent device, we probably want to change it. Most code of this size should run in 0-1ms (i.e. 1 is a rounding situation).

Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:168]: Yes, you're correct that in the event that the intent is coming from the default browser notification, it will have an affect on the startup time.

But this will only happen if the user taps on the notification.  Shouldn't we then perform user's action as soon as we can?

Comment (mcomella) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:168]: rocketsroger on Slack:

> I’m not sure if we should worry about the impact to performance when the intent is coming from the notification.  Since it only happens once in user’s lifetime and only happens if the reason for starting Fenix was because the user tapped on the notification.

Ah, it only happens once? Yeah, probably not worth optimizing then. We should probably add a comment to that effect. Good catch!

Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:168]: Sure, I'll add a comment  Thanks,

Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:168]: Added comment in `DefaultBrowserIntentProcessor.kt`

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt (hunk 7 above)
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt]: > Sure, I'll update the channel name to `Marketing` to allow for more generic usage in the future. Thanks,

Was this string something that the content/UX team verified? This will eventually be translated into various languages and deleting/updating this if we want to use channel grouping for something else will be annoying in the future.

**Comment by jonalmeida** [comment]
Let's make sure the notification configuration meets our general requirements. I'm not sure I've caught everything but some of it seems to be specific to the original "Powered by" notification which was intended to be a persistent notification to keep the browser running and notify the user that it was for PWAs, which is a different use case from this one.

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 9]
```diff
@@ -976,6 +986,26 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
...
+        val instanceWorkManager = WorkManager.getInstance(this)
+
+        if (!settings().shouldShowDefaultBrowserNotification()) {
+                // cancel notification work if already default browser
+                instanceWorkManager.cancelUniqueWork(DefaultBrowserNotificationWorker.NOTIFICATION_WORK_NAME)
+            return
+        }
+
+        val notificationWork = OneTimeWorkRequest.Builder(DefaultBrowserNotificationWorker::class.java)
+            .setInitialDelay(DefaultBrowserNotificationWorker.NOTIFICATION_DELAY, TimeUnit.MILLISECONDS)
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:999]: Ah, I think I can answer my own question, but you can confirm: this wouldn't happen because `defaultBrowserNotificationDisplayed` needs to be set to true and that only happens in the processor.

This was a bit confusing to follow because of the (nested) double negative. 😄

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 10]
```diff
@@ -488,7 +497,8 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
         )
 
         val intentProcessors =
-            listOf(CrashReporterIntentProcessor()) + externalSourceIntentProcessors
+            listOf(CrashReporterIntentProcessor()) +
+                    externalSourceIntentProcessors
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:501]: I'll fix it.  Thanks,

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 11]
```diff
@@ -976,6 +986,26 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
         }
     }
 
+    private fun setDefaultBrowserNotificationIfNeeded() {
+        val instanceWorkManager = WorkManager.getInstance(this)
+
+        if (!settings().shouldShowDefaultBrowserNotification()) {
+                // cancel notification work if already default browser
+                instanceWorkManager.cancelUniqueWork(DefaultBrowserNotificationWorker.NOTIFICATION_WORK_NAME)
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:994]: hmm yeah that's weird.  I'll fix as well.  Thanks,

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 12]
```diff
@@ -976,6 +986,26 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
         }
     }
 
+    private fun setDefaultBrowserNotificationIfNeeded() {
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:989]: Sure, I'll think about his

[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:989]: I'll move it into DefaultBrowserNotificationWorker

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+     * Build the default browser notification.
+     */
+    private fun buildNotification(): Notification {
+        val channelId = ensureChannelExists()
+        val intent = Intent(applicationContext, HomeActivity::class.java)
+        intent.putExtra(INTENT_DEFAULT_BROWSER_NOTIFICATION, true)
+
+        val pendingIntent = PendingIntent.getActivity(
+            applicationContext,
+            SharedIdsHelper.getNextIdForTag(applicationContext, NOTIFICATION_PENDING_INTENT_TAG),
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:46]: Yeah, `NOTIFICATION_PENDING_INTENT_TAG` is not really needed.  Just want to show what that value is.

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        with(applicationContext) {
+            val appName = getString(R.string.app_name)
+            return NotificationCompat.Builder(this, channelId)
+                .setSmallIcon(R.drawable.ic_status_logo)
+                .setContentTitle(
+                    applicationContext.getString(R.string.notification_default_browser_text, appName))
+                .setBadgeIconType(NotificationCompat.BADGE_ICON_SMALL)
+                .setColor(ContextCompat.getColor(this, R.color.primary_text_light_theme))
+                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
+                .setShowWhen(false)
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:60]: I kept it to false since we're not really showing a notification that is time sensitive.  I'm not sure which is better but personally I like the look of the notification without the timestamp.

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+                .setContentIntent(pendingIntent)
+                .setAutoCancel(true)
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:62]: This is something I added after testing.  Without `setAutoCancel(true)` the user will tap on the notification, set our browser to default and still see the notification is there.  Once the user tap on the notification it should be dismissed.

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt (hunk 7 above)
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt]: Yes, the original requirement does not have any text specified.  That's why I added @topotropic to review the text and she gave the approval above.

File: app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt (hunk 15 above)
[the remark this answers is not in the document]
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/onboarding/DefaultBrowserNotificationWorker.kt:67]: Sorry, I must have misread this when I reviewed it. Auto-cancel _true_ makes sense now - thanks!

**Review by jonalmeida — APPROVED** [review]
Looks good from eng side. ✅
